shared/js/ui/components/toggle.jsx

```diff
@@ -0,0 +1,203 @@
+import { useEffect, useLayoutEffect, useRef, useState } from 'preact/hooks'
+import { MDCSwitch } from '@material/switch'
+import { ns } from '../base/localize'
+import { isAndroid, isBrowser } from '../environment-check'
+
+/**
+ * @typedef {import('../templates/protection-header').ToggleState} ToggleState
+ * @typedef {import('../templates/protection-header').MigrationModel} MigrationModel
+ */
+
+/**
+ * @param {object} props
+ * @param {MigrationModel} props.model
+ */
+export function ProtectionToggle(props) {
+    const [toggleState, toggle] = useToggleState(props.model)
+    const altText = ns.site('updatingProtectionList.title')
+
+    return (
+        <div class={`site-info-toggle ${toggleState.active ? 'is-active' : ''}`}>
+            <p class="site-info__protection">
+                <span role="textbox" dangerouslySetInnerHTML={{ __html: toggleState.text }}></span>
+            </p>
+            <div class="site-info__toggle-container">
+                {toggleState.toggled && <img src="../img/spinner.svg" className="toggle-spinner" alt={altText} />}
+                {!toggleState.toggled && <ToggleButton toggleState={toggleState} onToggle={toggle} />}
+            </div>
+        </div>
+    )
+}
+
+/**
+ * @param {MigrationModel} model
+ * @returns {[ToggleState, () => void]}
+ */
+export function useToggleState(model) {
+    const [state, setState] = useState(() => {
+        const toggleState = {
+            text: ns.site('protectionsEnabled.title'),
+            active: true,
+            disabled: false,
+            label: '',
+            toggled: false,
+            sideEffects: false,
+        }
+
+        if (!model.protectionsEnabled) {
+            toggleState.text = ns.site('protectionsDisabled.title')
+            toggleState.active = false
+        }
+
+        /**
+         * Only the `extension` can toggle protections back ON (currently)
+         * So for all other platforms, a 'broken site' means that we disable the toggle
+         */
+        if (model.isBroken) {
+            if (!isBrowser()) {
+                toggleState.active = false
+                toggleState.disabled = true
+            }
+        }
+
+        const labelEnabled = ns.site('enableProtectionsSwitch.title')
+        const labelDisabled = ns.site('disableProtectionsSwitch.title')
+        toggleState.label = toggleState.active ? labelDisabled : labelEnabled
+
+        return toggleState
+    })
+
+    /**
+     * When the state changes, decide what to do
+     */
+    useEffect(() => {
+        if (!state.sideEffects) return
+
+        const int = setTimeout(() => {
+            // toggle, this communicates with native
+            model.toggleAllowlist()
+
+            // if the platform supports using a spinner
+            if (model.platformFeatures.spinnerFollowingProtectionsToggle) {
+                setState((prev) => {
+                    return { ...prev, toggled: true }
+                })
```

Comment:
❓ Any issue with potentially attempting to set state 300ms in the future when the component may have been unmounted?

Reply:
in this instance we have the same logic as we did in the old toggle - because we only allow it to be clicked once (which disables it). 

before it was done with a flag - but here we have `state.active` and `state.sideEffects` in the useEffect dependency array - so when they change any existing timers will be cancelled etc.